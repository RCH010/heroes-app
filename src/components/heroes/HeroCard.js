import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom';

const HeroCard = ({hero}) => {
    const {id, superhero, alter_ego, first_apperance, characters} = hero;
    return (
        <div className='card ms-3' style={{maxWidth: 540}}>
            <div className='row no-gutters'>
                <div className='col-md-4'>
                    <img src={`./assets/heroes/${id}.jpg`} className='card-img'alt={superhero}/>
                </div>
                <div className='col-md-8'>
                    <div className='card-body'>
                        <h2 className='card-title mb-0 h5'>
                            {superhero}
                        </h2>
                        <hr className='my-1'/>
                        <p className='card-text'>{alter_ego}</p>
                        {
                            (alter_ego !== characters) && <p className='card-text'>{characters}</p>
                        }
                        <p className='card-text'>
                            <small className='text-muted'>{first_apperance}</small>
                        </p>
                        <Link to={`./hero/${id}`}>
                            m√°s...
                        </Link>
                    </div>
                </div>
            </div>
        </div>
    )
}

HeroCard.propTypes = {
    hero: PropTypes.object.isRequired
}

export default HeroCard
