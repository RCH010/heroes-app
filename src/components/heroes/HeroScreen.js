import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { getHeroById } from '../../selectors/getHeroById';

export const HeroScreen = ({history}) => {
    const {heroeId} = useParams();

    // este use memo, hace que se memorice el resultado de la peticiÃ³n
    //   para no volver a hcer el proceso a menos que haya cambiado el param
    const heroInfo = useMemo(() => getHeroById(heroeId), [heroeId]);

    // Check hero exists
    if(!heroInfo) { return <Redirect to='/' />; }
    const {superhero, alter_ego, publisher, first_appearance, characters} = heroInfo;

    const handleReturn = () => {
        if(publisher === 'DC Comics') 
            history.push('./dc');
        else
            history.push('./');
    }
    
    return (
        <div className='row mt-5'>
            <div className='col-4'>
                <img 
                    alt={superhero} 
                    className='img-thumbnail animate__animated animate__fadeInLeft'
                    src={`../assets/heroes/${heroeId}.jpg`}  
                />
            </div>
            <div className='col-8'>
                <h1 className='h3'>{superhero}</h1>
                <ul className='list-group list-group-flush'>
                    <li className='list-group-item'><strong>Alter ego:</strong> {alter_ego}</li>
                    <li className='list-group-item'><strong>Publisher:</strong> {publisher}</li>
                    <li className='list-group-item'><strong>First Appearance:</strong> {first_appearance}</li>
                </ul>
                <h2 className='h4'>Characters</h2>
                <p className='px-3'>{characters}</p>
                <button 
                    className='btn btn-outline-info'
                    onClick={handleReturn}
                >
                    Return
                </button>
            </div>
        </div>
    )
}
