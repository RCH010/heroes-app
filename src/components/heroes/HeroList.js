import React, { useMemo } from 'react'
import PropTypes from 'prop-types'
import { getHeroByPublisher } from '../../selectors/getHeroByPublisher'
import HeroCard from './HeroCard';

export const HeroList = ({publisher}) => {
    // este use memo, hace que se memorice el resultado de la peticiÃ³n
    //   para no volver a hcer el proceso a menos que haya cambiado el param
    const heroes = useMemo(() => getHeroByPublisher(publisher), [publisher]);

    return (
        <div className='card-columns animate__animated animate__fadeIn'>
            {
                heroes.map( (hero) => (
                    <HeroCard key = {hero.id} hero= {hero} />
                ))
            }
        </div>
    )
}

HeroList.propTypes = {
    publisher: PropTypes.string.isRequired
}

export default HeroList
