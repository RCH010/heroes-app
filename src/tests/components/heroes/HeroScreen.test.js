import React from 'react';
import { mount } from "enzyme"
import { HeroScreen } from "../../../components/heroes/HeroScreen"
import { MemoryRouter, Route } from 'react-router-dom';

describe('Tests in <HeroScreen />', () => {
    
    const historyMock = {
        length: 10,
        goBack: jest.fn(),
        push: jest.fn(),
    };
    
    test('should display redirect component if there is not URL arguments', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={historyMock}/>
            </MemoryRouter>
        );
        expect(wrapper.find('Redirect').exists()).toBe(true);
    })
    
    test('should display correctly with an URL argument', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-iron']}>
                <Route path='/hero/:heroeId' component={HeroScreen} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should go back with push function to /', () => {
        const historyMock = {
            length: 1,
            goBack: jest.fn(),
            push: jest.fn(),
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-iron']}>
                <Route 
                    path='/hero/:heroeId' 
                    component={() => <HeroScreen history={historyMock}/>} 
                />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();
        expect(historyMock.push).toHaveBeenCalledWith('./');
    });

    test('should go back with to dc component', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/dc-arrow']}>
                <Route 
                    path='/hero/:heroeId' 
                    component={() => <HeroScreen history={historyMock}/>} 
                />
            </MemoryRouter>
        );
        wrapper.find('button').prop('onClick')();
        expect(historyMock.push).toHaveBeenCalledWith('./dc');
    })
    
    test('should call redirect if heroId doesnt exists', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/dc-arrow12122']}>
                <Route 
                    path='/hero/:heroeId' 
                    component={() => <HeroScreen history={historyMock}/>} 
                />
            </MemoryRouter>
        );

        expect(wrapper.text()).toBe('');
    })
    
})
