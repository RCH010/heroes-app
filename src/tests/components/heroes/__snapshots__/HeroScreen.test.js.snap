// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <HeroScreen /> should display correctly with an URL argument 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Iron Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="../assets/heroes/marvel-iron.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <h1
      className="h3"
    >
      Iron Man
    </h1>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <strong>
          Alter ego:
        </strong>
         
        Tony Stark
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          Publisher:
        </strong>
         
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          First Appearance:
        </strong>
         
        Tales of Suspense #39
      </li>
    </ul>
    <h2
      className="h4"
    >
      Characters
    </h2>
    <p
      className="px-3"
    >
      Tony Stark
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
